### PR TITLE
change(android,web) Use web-based popup key longpresses

### LINKED ISSUE #9499: bug(android): Special character '<' rendering problem with IPA (SIL) keyboard
### Describe the bug

I noticed that the special character "<" is missing in the long press option while using IPA (SIL) keyboard. 

### Reproduce the bug

1. Install Keyman 17.0.162-alpha build. 
2. Open Keyman In-App.
3. Download and Install IPA (SIL) Keyboard. 
4. Switch to IPA (SIL) keyboard in the Keyman App. 
5. Click the Number (123) key. 
6. Long Press "]" key. 
7. From the Longpress options, Select "<" special character. 

Here, I noticed that the "<" symbol does not appear on the screen. Other two special characters (>>, }) appear correctly. Seems to be an issue. 

I have attached the Screenshot as well as the Video file for reference. 

.. special character is missing

   

https://github.com/keymanapp/keyman/assets/19683143/a0e17149-2848-4f5c-99f8-de748791bac9

### Expected behavior

Clicking the "<" special character on the longpress option should be correctly and rendered and appear on the text screen. 

### Related issues

_No response_

### Keyman apps

- [X] Keyman for Android
- [ ] Keyman for iPhone and iPad
- [ ] Keyman for Linux
- [ ] Keyman for macOS
- [ ] Keyman for Windows
- [ ] Keyman Developer
- [ ] KeymanWeb
- [ ] Other - give details at bottom of form

### Keyman version

17.0.162-alpha

### Operating system

Android 5.0, 9.0, 12.0

### Device

Android 12.0 /  API 31 emulator, Redmi Note 8 Pro

### Target application

_No response_

### Browser

_No response_

### Keyboard name

IPA (SIL)

### Keyboard version

1.8.6

### Language name

und-Latn

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by darcywong00**
Addresses part of #9573 and fixes #9499

This updates Keyman Engine for Android to match iOS using KeymanWeb for handling longpresses.
The longpress keys will be limited by the top of the banner/OSK.

### KeymanWeb changes ###
* Remove Android-specific portions of `buildEmbeddedGestureConfig` and `pendingLongpress`

### Keyman Engine for Android changes ###
* Remove native Java code for popup, preview, and longpresses keys, as it's handled in KeymanWeb.

> Note: this means banner will always be visible on Android, no option to switch off banner any more.

This portion of #9573 will be addressed on a separate PR. Android will possibly need some image banners to display when predictions are disabled.

## User Testing
**Setup** - Install the PR build of Keyman for Android on an Android device emulator. In the Keyman app, also install the IPA (SIL) keyboard

* **TEST_IPA_RENDERING** - Verifies #9499 is fixed
1. Install the PR build of Keyman for Android on an Android device/emulator
2. Open Keyman In-App and dismiss the "Get Started" menu
3. Switch to IPA (SIL) keyboard in the Keyman App
4. Click the Number (123) key
5. Long press <kbd>[</kbd> key
6. From the longpress options, select <kbd><</kbd> special character
7. Verify the `<` symbol appears on the screen.

* **TEST_ANDROID_INAPP** - Verifies longpresses and globe key work in the Keyman app
1. Install the PR build of Keyman for Android on an Android device/emulator
2. Open Keyman In-App and select sil_eurolatin keyboard
3. On the default sil_eurolatin keyboard, briefly press a top row key and release
4. Verify the key preview briefly appears, and the selected key is output
5. Longpress on a top row key and hold
6. Verify the longpress keys appear (position limited to the top of the banner)
7. Select a longpress key and release
8. Verify the selected key is output
9. Short-press the globe key
10. Verify the Keyman keyboard switches to IPA (SIL) keyboard
11. Long-press the globe key and release
12. Verify the keyboard picker menu appears.
13. Select a keyboard
14. Verify Keyman switches to the selected keyboard

* **TEST_ANDROID_SYSTEM_LONGPRESSES** - Verifies longpresses and globe key work as a system keyboard
1. Install the PR build of Keyman for Android on an Android device/emulator
2. Open Keyman In-App and select the "Get Started" menu
3. Enable Keyman as a system keyboard and set as default keyboard
4. Launch a separate app (e.g. Chrome browser) and select a text area
5. With default Keyman sil_eurolatin keyboard, briefly press a top row key and release
6. Verify the key preview briefly appears, and the selected key is output
7. Longpress on a top row key and hold
8. Verify the longpress keys appear (position limited to the top of the banner)
9. Select a longpress key and release
10. Verify the selected key is output
11. Short-press the globe key
12. Verify the Keyman keyboard switches to IPA (SIL) keyboard
13. Long-press the globe key and release
14. Verify the keyboard picker menu appears.
15. Select a keyboard
16. Verify Keyman switches to the selected keyboard

* **TEST_IOS_INAPP** - Verifies longpresses key works in the Keyman for iPhone and iPad app
1. Install the PR build of Keyman for iPhone and iPad on an iOS device/emulator
2. Open Keyman In-App and select sil_eurolatin keyboard
3. On the default sil_eurolatin keyboard, briefly press a top row key and release
4. Verify the key preview briefly appears, and the selected key is output
5. Longpress on a top row key and hold
6. Verify the longpress keys appear (position limited to the top of the banner)
7. Select a longpress key and release
8. Verify the selected key is output

### CODE REVIEW

File: web/src/resources/osk/kmwosk.css
Comment (darcywong00) [inline web/src/resources/osk/kmwosk.css:493]: Longpress background styling to match the current background from
https://github.com/keymanapp/keyman/blob/31fd6ea42afef1dddd7d883fc31c5086fe9ae9b2/android/KMEA/app/src/main/res/values/colors.xml#L11

File: web/src/app/webview/src/osk/pendingLongpress.ts
Comment (jahorton) [inline web/src/app/webview/src/osk/pendingLongpress.ts:41]: `SubkeyDelegator` may be removed as well; it's used exclusively as a follow-up to this class.

File: web/src/app/webview/src/keymanEngine.ts
Comment (jahorton) [inline web/src/app/webview/src/keymanEngine.ts:72]: - [ ] Undo this change.

Your changes in `oskConfiguration.ts`, which remove the Android-specific parts of the function being called here, should be sufficient.

This change _additionally_ removes the globe-key hint message data that allows the iOS app to provide a hint about the keyboard-change key.  I'm not sure if Android has that as well; if it does, Android also uses the same thing.  (See lines 43-47 of `oskConfiguration.ts`.)

While not the exact matching commit, this is the corresponding globe-hint section to be restored:

https://github.com/keymanapp/keyman/blob/e7dedfc8ee46eba90ff3c725c4987eba8d8f1b32/web/src/app/webview/src/oskConfiguration.ts#L42-L46

**Review by jahorton — CHANGES_REQUESTED** [review]

File: web/src/app/webview/src/osk/pendingLongpress.ts
Comment (darcywong00) [reply web/src/app/webview/src/osk/pendingLongpress.ts:41]: ah, I still had to keep `SubkeyDelegator` for
https://github.com/keymanapp/keyman/blob/9fbe5f6b410df501070d95b31143ffc253fa70f7/web/src/app/webview/src/keymanEngine.ts#L109

Comment (jahorton) [reply web/src/app/webview/src/osk/pendingLongpress.ts:41]: Right, which is within the `executePopupKey` method.  Which is no longer being called, correct?

File: android/KMEA/app/src/main/assets/android-host.js
Comment (jahorton) [inline android/KMEA/app/src/main/assets/android-host.js:328]: Re: a recent comment in one of the other threads.

File: web/src/app/webview/src/osk/pendingLongpress.ts
Comment (darcywong00) [reply web/src/app/webview/src/osk/pendingLongpress.ts:41]: I thought both iOS and Android will still need KeymanWeb to handle `executePopupKey`.

**Comment by darcywong00** [comment]
@bharanidharanj - please hold off on user testing until reviewers approve changes. Thanks

File: web/src/app/webview/src/osk/pendingLongpress.ts
Comment (darcywong00) [reply web/src/app/webview/src/osk/pendingLongpress.ts:41]: gotcha. I was able to remove popVisible (it had the same gesture)

Comment (jahorton) [reply web/src/app/webview/src/osk/pendingLongpress.ts:41]: > I thought both iOS and Android will still need KeymanWeb to handle `executePopupKey`.

Ah, I get that.  Yeah, we did leave the Swift connections to it in place, since back when we moved iOS to rely on the WebView, we were hoping Apple would fix the motivating bug.  Didn't happen, though, and we never revisited to clean out the old code.  Probably worth making into a cleanup issue.

**Review by jahorton — APPROVED** [review]
I'm satisfied with the fixes to my requests for changes at this point.

File: android/KMEA/app/src/main/java/com/keyman/engine/KMKeyboard.java
Comment (mcdurdin) [inline android/KMEA/app/src/main/java/com/keyman/engine/KMKeyboard.java:368]:
```suggestion
    gestureDetector.onTouchEvent(event);
```

But do we even need `gestureDetector` any more?

File: web/src/app/webview/src/oskConfiguration.ts
Comment (mcdurdin) [inline web/src/app/webview/src/oskConfiguration.ts:50]: `createKeyTip` and `startLongpress` should be removed from `EmbeddedGestureConfig`.

I also think that the globe hint should be reimplemented in KeymanWeb, so we don't need any of this glue, and so it can be themed later. (This can be a separate pull)

**Review by mcdurdin — APPROVED** [review]
Changes LGTM. Minor suggestions.

File: web/src/app/webview/src/oskConfiguration.ts
Comment (darcywong00) [reply web/src/app/webview/src/oskConfiguration.ts:50]: `createKeyTip` is used in `oskView` and `visualKeyboard`
`startLongpress` is used in `visualKeyboard`

@jahorton good with removing these?

----
globe hint split into separate issue #9596

Comment (darcywong00) [reply web/src/app/webview/src/oskConfiguration.ts:50]: > `createKeyTip` and `startLongpress` should be removed from `EmbeddedGestureConfig`.

`EmbeddedGestureConfig` cleanup in 98660fef562ed2eed904e53b5fca568928a5f74f

**Comment by bharanidharanj** [comment]
* **TEST_ANDROID_SYSTEM_LONGPRESSES (PASSED):** Tested in Android 12.0 / API31 emulator and here is my observation: 1. Installed the attached PR build 17.0.176-alpha-test-9591 on an Android emulator. 2. Verified that the long press key on the top row is working as expected for sil_eurolatin keyboard. 3. Verified that the keyboard has been changed to IPA (SIL) after short-press and releasing the globe key. 4. Also, verified that the keyboard picker menu appears, after long press and release of the globe key. Able to switch to some other keyboard from the Keyboard picker menu. Seems to be working as expected.

**Comment by bharanidharanj** [comment]
* **TEST_ANDROID_SYSTEM_LONGPRESSES (PASSED):** Tested in Android 12.0 / API31 emulator and here is my observation: 1. Installed the attached PR build 17.0.176-alpha-test-9591 on an Android emulator. 2. Enabled Keyman as the default keyboard using Get started dialog. Opened Chrome browser. Verified that the long press key on the top row is working as expected for the sil_eurolatin keyboard. 3. Verified that the keyboard has been changed to IPA (SIL) after short-press and releasing the globe key. 4. Also, verified that the keyboard picker menu appears, after long press and release of the globe key. Able to switch to some other keyboard from the Keyboard picker menu. Seems to be working as expected.

**Comment by bharanidharanj** [comment]
* **TEST_IPA_RENDERING (PASSED):** Tested in Android 12.0 / API31 emulator and here is my observation: 1. Installed the attached PR build 17.0.176-alpha-test-9591 on an Android emulator. 2. Switched to IPA (SIL) keyboard in the Keyman In-App. 3. Verified that the < symbol appears on the text input screen after selecting it from the long-press options. Seems to be working as expected.

**Comment by bharanidharanj** [comment]
# Test Results

* **TEST_IOS_INAPP (PASSED):** Tested in iPhone 13 Pro Max / iOS 15.0 simulator and here is my observation: 1. Installed the attached PR build 17.0.176-alpha-test-9591 on iPhone Simulator. 2. Switched to sil_eurolatin keyboard in the Keyman In-App. 3. Verified that the key preview briefly appears after briefly pressing a top row key and release. 4. Verified that the selected key appeared on the screen after selecting it from the longpress options. Seems to be working as expected.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 17.0.179-alpha](https://keyman.com/downloads/releases/alpha/17.0.179)

**Pull request merged**
